### PR TITLE
snowflake IO manager and pandas type handler

### PR DESCRIPTION
**Pull request by sryza**
### Summary & Motivation

Snowflake is probably the most common data warehouse among our users, but the current integration we provide for it doesn't fit our preferred pattern of sequestering IO inside IO managers.  This has come up with recent questions on Slack a few times.

Snowflake tables can correspond to multiple in-memory formats. E.g. Pandas DataFrames, Spark DataFrames. This PR proposes a pluggable interface for in-memory formats that allows users to only load the dependencies for the in-memory formats / processing frameworks that they use.

This PR includes the Snowflake IO manager and a pandas type handler implementation. It's possible that the underlying abstractions like TableSlice and DbIOManager could be used for integrating with other databases as well, but for now they're kept private to limit surface area and allow iteration.

Followups to this PR would include:
* A Spark type handler
* dbt
* Using this in the hacker news demo

### How I Tested These Changes

Unit tests.

### CODE REVIEW

File: python_modules/libraries/dagster-snowflake/dagster_snowflake_tests/test_db_io_manager.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    output_context = build_output_context(
+        asset_key=AssetKey(["schema1", "table1"]), resource_config=resource_config
+    )
+    manager.handle_output(output_context, 5)
+    input_context = build_input_context(
+        upstream_output=output_context,
+        resource_config=resource_config,
+        dagster_type=resolve_dagster_type(int),
+    )
+    assert manager.load_input(input_context) == 7
```
Comment (gibsondan) [inline python_modules/libraries/dagster-snowflake/dagster_snowflake_tests/test_db_io_manager.py:65]: could const this

File: python_modules/libraries/dagster-snowflake/dagster_snowflake_tests/test_db_io_manager.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    table_slice = TableSlice(database="database_abc", schema="schema1", table="table1")
+    assert int_handler.handle_output_calls[0][1:] == (table_slice, 5)
+    db_client.delete_table_slice.assert_called_once_with(output_context, table_slice)
+
+    input_context = build_input_context(
+        upstream_output=output_context,
+        resource_config=resource_config,
+        dagster_type=resolve_dagster_type(str),
+    )
+    assert manager.load_input(input_context) == "8"
```
Comment (gibsondan) [inline python_modules/libraries/dagster-snowflake/dagster_snowflake_tests/test_db_io_manager.py:98]: this too

**Review by gibsondan — APPROVED** [review]

**Review by dpeng817 — COMMENTED** [review]
This is a pretty cool pattern, and we should probably adopt it across all of our io managers. To make it a bit more accessible, is there a dagster-default set of type handlers that you think would be feasible to add? Or at least the tools to construct them?

**Review by OwenKephart — COMMENTED** [review]
big fan of this addition, although I'm somewhat worried about putting this in its own separate library, as it implies a pattern where we'll end up with some cross product set of libraries mapping between each storage location (snowflake, redshift, postgres, etc.) and each dataframe medium (pandas, pyspark).

Is the concern just that adding this to dagster-snowflake would add pandas as a dependency to that library (when often people will not need that dep?). If so, we could avoid that by having this integration as an "extra", and have people install it with something like

`pip install dagster-snowflake[pandas]`

**Comment by prha** [comment]
+1 to @OwenKephart 's comments

**Comment by sryza** [comment]
The `extra_requires` vs. "separate Python package" debate is one that I believe has been going on since the founding of Elementl.

I have heard @schrockn voice strong opinions in favor of the "separate Python packages" approach, and I am less sure but lean towards that approach as well.  His argument, if I'm representing correctly, is that `extra_requires` is basically a shadow dependency management system that makes dependencies less explicit and less easy to reason about.

### Concrete situation: `extra_requires` approach

Step 1: you copy / paste a code snippet like this:
```
from dagster_snowflake import make_snowflake_io_manager, SnowflakePandasTypeHandler
...
```

Step 2: You mistakenly pip install "dagster-snowflake", even though you should have installed "dagster-snowflake[pandas]"

Step 3: You run your code. You hit "ModuleNotFoundError: No module named 'pandas'" or maybe "snowflake has no attribute "sqlalchemy" (because the snowflake pandas type handler depends on the snowflake-sqlalchemy integration).

This is confusing, because the errors you're hitting don't directly correspond to your pip install mistake.

### Concrete situation: "separate Python package" approach
Step 1: you copy / paste a code snippet like this:
```
from dagster_snowflake import make_snowflake_io_manager
from dagster_snowflake_pandas import SnowflakePandasTypeHandler
...
```

Step 2: You pip install "dagster-snowflake" but forget to install "dagster-snowflake-pandas".

Step 3: You run your code. You hit "ModuleNotFoundError: No module named 'dagster_snowflake_pandas'".

### Separate argument in favor of the "separate Python package" approach

It encourages users to roll their own integrations and host them themselves. If the Pandas and PySpark integrations both live inside the dagster-snowflake package, users who want to use some other special dataframe library like Dask will be more likely to feel like that needs to live inside the dagster-snowflake package too.

**Comment by sryza** [comment]
> To make it a bit more accessible, is there a dagster-default set of type handlers that you think would be feasible to add? Or at least the tools to construct them?

@dpeng817 mind expanding a little bit on what you mean here?

**Comment by OwenKephart** [comment]
I'll admit that my past experiences with `pip install foo_package[bar]` have been plagued by similar issues. The reasoning makes sense :+1:

**Review by OwenKephart — APPROVED** [review]
Super excited about this -- among a bunch of other benefits, supplying this capability out of the box is going to make writing example code way easier.

File: python_modules/libraries/dagster-snowflake/dagster_snowflake/db_io_manager.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+class TableSlice(NamedTuple):
+    table: str
+    schema: str
+    database: str
+    columns: Optional[Sequence[str]] = None
+    partition: Optional[TablePartition] = None
+
+
+class DbTypeHandler(ABC, Generic[T]):
```
[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/libraries/dagster-snowflake/dagster_snowflake/db_io_manager.py:38]: You should be able to accomplish that under the current regime, by registering multiple DbTypeHandlers with your IO manager and then using different types on the inputs and outputs.

**Pull request merged**